Change AbstractExpressions to raw pointers

GustavoAngulo opened the pull request:
Fixes #384 

This PR changes how we store `AbstractExpression` in the system. Previously, we used `std::shared_ptr` for `AbstractExpression`. However, because shared pointers do not provide explicit ownership over the pointer, it became difficult to transactionally store `AbstractExpression` in the catalog particularly. Thus, the decision was made to change these to raw pointers to allow more fine grained ownership and deallocation.

GustavoAngulo commented:
Need to revert an accidental change actually. Should be ready for review soon

jrolli commented on src/include/parser/postgresparser.h:
```diff
@@ -172,7 +172,7 @@ class PostgresParser {
   // INSERT helpers
   static std::unique_ptr<std::vector<std::string>> ColumnNameTransform(List *root);
-  static std::unique_ptr<std::vector<std::vector<std::shared_ptr<AbstractExpression>>>> ValueListsTransform(List *root);
+  static std::unique_ptr<std::vector<std::vector<AbstractExpression *>>> ValueListsTransform(List *root);
```
Why are we returning a unique vector (which is itself a pointer to an array) rather than returning the vector?

GustavoAngulo replied:
Similarly, responded above.

GustavoAngulo commented on src/include/common/json.h:
```diff
@@ -11,15 +11,27 @@ using json = nlohmann::json;
+      j = nullptr;                                                                             \
+    }                                                                                          \
+  }                                                                                            \
+  inline void to_json(nlohmann::json &j, const std::shared_ptr<ClassName> c) { /* NOLINT */    \
```
Yes, the plan nodes are still `std::shared_ptr`, and this interface is used during their serialization ([example](https://github.com/cmu-db/terrier/blob/3d88f8ca80f6c3804a246fbca046b3ab08305391/src/planner/plannodes/abstract_plan_node.cpp#L40)). Like you said, if we want to consider changing how we pass plan nodes around, that would be outside this PR's scope.

GustavoAngulo commented on src/include/parser/expression/abstract_expression.h:
```diff
@@ -77,11 +81,11 @@ class AbstractExpression {
 
   /**
-   * Creates a (shallow) copy of the current AbstractExpression.
+   * Creates a (deep) copy of the current AbstractExpression.
```
I think you're right. In that case, I don't think we need `Copy()` at all, if we want to enforce doing shallow copies, and forcing manual deep copies, we might be better off marking the copy constructor/assignment as `default`, and deleting the move constructor/assignment.

GustavoAngulo commented on src/include/parser/delete_statement.h:
```diff
@@ -48,13 +48,13 @@ class DeleteStatement : public SQLStatement {
  private:
   std::shared_ptr<TableRef> table_ref_;
-  std::shared_ptr<AbstractExpression> expr_;
+  AbstractExpression *expr_;
```
Yea it should be, I forgot to do a pass on this.

GustavoAngulo commented on src/include/parser/insert_statement.h:
```diff
@@ -38,7 +38,7 @@ class InsertStatement : public SQLStatement {
    */
   InsertStatement(std::shared_ptr<std::vector<std::string>> columns, std::shared_ptr<TableRef> table_ref,
-                  std::shared_ptr<std::vector<std::vector<std::shared_ptr<AbstractExpression>>>> insert_values)
+                  std::shared_ptr<std::vector<std::vector<AbstractExpression *>>> insert_values)
```
This looks like just bad code, possibly a remnant of peloton. I've removed it. These expressions are converted to transient values during the planning phase, so the lifecycle does not merit a complicated way to store the values.

GustavoAngulo commented on src/include/parser/insert_statement.h:
```diff
@@ -77,15 +85,15 @@ class InsertStatement : public SQLStatement {
   const std::shared_ptr<SelectStatement> select_;
   // TODO(WAN): unsure about this one.
-  const std::shared_ptr<std::vector<std::vector<std::shared_ptr<AbstractExpression>>>> insert_values_;
+  const std::shared_ptr<std::vector<std::vector<AbstractExpression *>>> insert_values_;
```
Responded above.

GustavoAngulo commented on src/include/parser/expression/subquery_expression.h:
```diff
@@ -26,10 +26,12 @@ class SubqueryExpression : public AbstractExpression {
     // TODO(WAN): Previous codebase described as a hack, will we need a deep copy?
     // Tianyu: No need for deep copy if your objects are always immutable! (why even copy at all, but that's beyond me)
-    return std::make_shared<SubqueryExpression>(*this);
+    return new SubqueryExpression(std::make_shared<parser::SelectStatement>(*subselect_));
```
Fixed.

GustavoAngulo commented on src/include/parser/expression/aggregate_expression.h:
```diff
@@ -20,15 +20,23 @@ class AggregateExpression : public AbstractExpression {
    */
   AggregateExpression() = default;
 
-  std::shared_ptr<AbstractExpression> Copy() const override { return std::make_shared<AggregateExpression>(*this); }
```
Yea you're probably right. We def need to have a discussion on smart pointers as a group

GustavoAngulo commented:
#394 Added `ManagedPointer` which (fingers crossed) fixes our fear of raw pointer lifecycle. The catalog does not need to worry about this however, as all `AbstractExpression` it creates should be managed by other components (catalog, executor, optimizer). It is up to those components to wrap the raw pointers created by the parser using `ManagedPointer` if they need to control its lifecycle. Thus, there is no need to change anything in this PR after the addition of `ManagedPointer`.  

As for the `Copy()` method, because we are now using raw pointers, shallow copies are easy, so I've made `Copy()` perform deep copies for when it's needed, as well as included a warning that deep copies should be used sparingly.

jrolli commented:
Should the raw pointers in the statements be `ManagedPtr` now that we have that concept?

GustavoAngulo commented:
> Should the raw pointers in the statements be `ManagedPtr` now that we have that concept?

I think statements should store them as raw pointers, since if they store them as managed pointers they won't be able to delete them themselves. What I can do is change the getter methods to return `ManagedPointer`, I think thats how we envision `ManagedPointer` being passed around.

GustavoAngulo commented:
Per today's meeting, I've changed getters to return `ManagedPointer`. I also added a [`CastManagedPointer()`](https://github.com/cmu-db/terrier/pull/386/files#diff-9284e61bbf98382241fe4e6b842bd8d9R81) method that allows down casting the underlying pointer of a `ManagedPointer`.

jrolli commented on src/include/planner/plannodes/create_table_plan_node.h:
```diff
@@ -607,8 +607,8 @@ class CreateTablePlanNode : public AbstractPlanNode {
-        parser::ConstantValueExpression *const_expr_elem =
-            dynamic_cast<parser::ConstantValueExpression *>(col->GetCheckExpression()->GetChild(1).get());
+        auto const_expr_elem =
+            col->GetCheckExpression()->GetChild(1).CastManagedPointerTo<const parser::ConstantValueExpression>();
```
I'm pretty sure this incorporates two undocumented and potentially invalid assumptions:
1. This is assumes the check expression has at least two children (not enforced anywhere).  My guess is that someone is assuming that it will always be an operator.  However, what if it is a UDF?
2. This assumes that the second argument must be constant.  I'm pretty sure column constraints can be arbitrary boolean expressions.

At a minimum the assumptions and their justifications need to be documented, but I think the real solution is probably storing the expression in `check_info` and not transient values.  I've been working on the assumption that constraints would be described by expressions as I write the catalog for the same reason that default values and index key descriptions are expressions and not transient values.

jrolli commented on test/parser/parser_test.cpp:
```diff
@@ -184,19 +184,19 @@ TEST_F(ParserTestBase, CreateIndexTest) {
+  auto ia1 = create_stmt->GetIndexAttributes()[0]->GetExpression();
   EXPECT_EQ(ia1->GetExpressionType(), ExpressionType::OPERATOR_MINUS);
-  auto ia1l = reinterpret_cast<TupleValueExpression *>(ia1->GetChild(0).get());
+  auto ia1l = ia1->GetChild(0).CastManagedPointerTo<const TupleValueExpression>();
```
I'll concede that `CastManagedPointerTo` is necessary/useful for the parser test, but I'm not convinced its needed in production code.  Should we make it a private function with this test as a "TEST_FRIEND"?

apavlo commented:
@lmwnshn we need your expertise on this...

mush-zhang commented:
The only usage of the .get() method in ManagedPointer is in parser test of update clause. As you have already changed `value` of updateClause to raw pointer, the managedPointer::get() is essentially used nowhere. So probably get rid of ManagedPointer::get() method

lmwnshn commented:
Jumping ship to #483, with a narrowed scope of just expressions.
